### ISSUE TITLE
Consider all related snapshots for incremental send/receive (preventing duplication of data)

### ISSUE THREAD

**Issue by fturco** [digint/btrbk#255]
I'm using btrbk-0.26.1 on a Gentoo Linux system.

I have a single btrfs filesystem with 4 subvolumes:
- root-i3
- home-i3
- root-plasma
- home-plasma

Basically I'm not sure if I should use the i3 window manager or the KDE Plasma desktop environment, so I keep both of them on my system until make up my mind.

I created the `root-i3` and the `home-i3` subvolumes after booting my KDE Plasma system and running the following commands:

    btrfs subvolume snapshot / /mnt/toplevel/root-i3
    btrfs subvolume snapshot /home /mnt/toplevel/home-i3

The previous commands took just a few seconds and almost no additional space on disk was used.

After that, I booted the i3 subvolumes and replaced KDE Plasma with i3.

I'm now trying to back up my entire btrfs filesystem (both i3 and KDE Plasma) to an external hard disk with btrbk. The problem is that the target backups are taking approximately twice the disk space of the source. For example since my home directory uses about 1 TiB of space, btrbk is copying approximately 2 TiB of data to my external hard disk. Instead, I was expecting it to copy 1 TiB plus a few additional GiB for minor differences between `home-i3` and `home-plasma`. In other words, it seems that it btrbk is not aware `home-i3` and `home-plasma` are closely related and is duplicating them instead.

How can I fix this problem?

I'm using the following btrbk configuration file:

```
snapshot_dir snapshots
timestamp_format long

volume /mnt/toplevel

  subvolume root-i3
  target send-receive /media/samsung-red

  subvolume home-i3
  target send-receive /media/samsung-red

  subvolume root-plasma
  target send-receive /media/samsung-red

  subvolume home-plasma
  target send-receive /media/samsung-red
```

**Comment by fturco** [comment]
```
# ./btrbk dryrun -t --config /etc/btrbk/samsung-red.conf 
TRANSACTION LOG
type          status          target_host  target_subvol                                  source_host  source_subvol                                  parent_subvol
------------  --------------  -----------  ---------------------------------------------  -----------  ---------------------------------------------  ---------------------------------------------
startup       v0.27.0         -            -                                              -            -                                              -
snapshot      dryrun_success  -            /mnt/toplevel/snapshots/i3.20181018T1954       -            /mnt/toplevel/i3                               -
snapshot      dryrun_success  -            /mnt/toplevel/snapshots/plasma.20181018T1954   -            /mnt/toplevel/plasma                           -
snapshot      dryrun_success  -            /mnt/toplevel/snapshots/storage.20181018T1954  -            /mnt/toplevel/storage                          -
send-receive  dryrun_success  -            /media/samsung-red/i3.20181018T1954            -            /mnt/toplevel/snapshots/i3.20181018T1954       /mnt/toplevel/snapshots/i3.20181018T1333
send-receive  dryrun_success  -            /media/samsung-red/plasma.20181018T1954        -            /mnt/toplevel/snapshots/plasma.20181018T1954   /mnt/toplevel/snapshots/plasma.20181018T1333
send-receive  dryrun_success  -            /media/samsung-red/storage.20181018T1954       -            /mnt/toplevel/snapshots/storage.20181018T1954  /mnt/toplevel/snapshots/storage.20181018T1333
finished      success         -
```

Please note that I made some changes since my previous message:
- The `root-i3` subvolume has been renamed to `i3`
- The `root-plasma` subvolume has been renamed to `plasma`
- The `home-plasma` subvolume has been deleted
- The `home-i3` subvolume has been renamed to `storage`

But `i3` should still have `plasma` as its parent. I hope renaming subvolumes doesn't break parent-children relationships...

**Comment by digint** [comment]
Hmm, looks like you already transferred both `i3.20181018T1333` and `plasma.20181018T1333` earlier? If you delete `plasma.*` on target, btrbk should use `i3.20181018T1954` as parent_subvol on the next run, saving 1TB.

> I hope renaming subvolumes doesn't break parent-children relationships...

No, renaming is not a problem. The relationships are based on the "Parent UUID" (e.g. `btrfs sub show /mnt/toplevel/snapshots/i3.20181018T1954`). These do not change when renaming.

**Comment by fturco** [comment]
OK. I deleted the `plasma*` subvolumes from the target and run `btrbk` from the `find-all-related` branch again. It seems to work! Instead of copying more than 100 GiB of data, it only copied less than 20 GiB. Thank you!

**Comment by digint** [comment]
This one has finally made it to master:

- cb23c65 btrbk: get_related_subvolumes: add direct children from parent chain
- d64e237 btrbk: get_best_parent: consider all parent/child relations
- ( 514e692 btrbk: add "incremental_resolve" configuration option )

**Comment by digint** [comment]
merged in btrbk-0.28.0, closing issue
